staging/vhost-device-gpu/src/lib.rs

```diff
@@ -42,58 +42,260 @@ pub mod protocol;
+            }
+            first = false;
+
+            match capset {
+                Self::VIRGL => write!(f, "virgl"),
+                Self::VIRGL2 => write!(f, "virgl2"),
+                Self::GFXSTREAM_VULKAN => write!(f, "gfxstream-vulkan"),
+                Self::VENUS => write!(f, "venus"),
+                Self::DRM => write!(f, "drm"),
+                Self::GFXSTREAM_MAGMA => write!(f, "gfxstream-magma"),
+                Self::GFXSTREAM_GLES => write!(f, "gfxstream-gles"),
+                Self::GFXSTREAM_COMPOSER => write!(f, "gfxstream-composer"),
+                _ => panic!("Unknown capset {:#x}", self.bits()),
+            }?;
+        }
+
+        Ok(())
+    }
+}
+
+impl GpuCapset {
+    /// Return the number of enabled capsets
+    pub const fn num_capsets(self) -> u32 {
+        self.bits().count_ones()
+    }
+}
+
 #[derive(Debug, Clone)]
-/// This structure holds the internal configuration for the GPU backend,
-/// derived from the command-line arguments provided through `GpuArgs`.
+/// This structure holds the configuration for the GPU backend
 pub struct GpuConfig {
-    /// vhost-user Unix domain socket
-    socket_path: PathBuf,
     gpu_mode: GpuMode,
+    capset_mask: GpuCapset,
+    flags: GpuFlags,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct GpuFlags {
+    pub use_egl: bool,
+    pub use_glx: bool,
+    pub use_gles: bool,
+    pub use_surfaceless: bool,
+}
+
+impl GpuFlags {
+    // `const` version of `default()`
+    pub const fn new_default() -> Self {
+        Self {
+            use_egl: true,
+            use_glx: false,
+            use_gles: true,
+            use_surfaceless: true,
+        }
+    }
+}
+
+impl Default for GpuFlags {
+    fn default() -> Self {
+        Self::new_default()
+    }
+}
+
+#[derive(Debug, ThisError)]
+pub enum GpuConfigError {
+    #[error("The mode {0} does not support {1} capset")]
+    CapsetUnsuportedByMode(GpuMode, GpuCapset),
+    #[error("Requested gfxstream-gles capset, but gles is disabled")]
+    GlesRequiredByGfxstream,
 }
 
 impl GpuConfig {
+    const DEFAULT_VIRGLRENDER_CAPSET_MASK: GpuCapset = GpuCapset::VIRGL
+        .union(GpuCapset::VIRGL2)
+        .union(GpuCapset::VENUS);
+    const DEFAULT_GFXSTREAM_CAPSET_MASK: GpuCapset = GpuCapset::GFXSTREAM_VULKAN;
+
     /// Create a new instance of the `GpuConfig` struct, containing the
     /// parameters to be fed into the gpu-backend server.
-    pub const fn new(socket_path: PathBuf, gpu_mode: GpuMode) -> Self {
-        Self {
-            socket_path,
-            gpu_mode,
+    pub fn new(
+        gpu_mode: GpuMode,
+        capset: Option<GpuCapset>,
+        flags: GpuFlags,
+    ) -> Result<Self, GpuConfigError> {
+        let capset = match (gpu_mode, capset) {
+            (GpuMode::VirglRenderer, None) => Self::DEFAULT_VIRGLRENDER_CAPSET_MASK,
+            (GpuMode::Gfxstream, None) => Self::DEFAULT_GFXSTREAM_CAPSET_MASK,
+            (_, Some(capset_mask)) => capset_mask,
+        };
```

Comment:
can we abstract this part into a `get_default_capset helper`? 
```suggestion
    fn get_default_capset(gpu_mode: GpuMode, capset: Option<GpuCapset>) -> Result<GpuCapset, GpuConfigError> {
        match capset {
            Some(capset_mask) => Ok(capset_mask),
            None => match gpu_mode {
                GpuMode::VirglRenderer => Ok(Self::DEFAULT_VIRGLRENDER_CAPSET_MASK),
                GpuMode::Gfxstream => Ok(Self::DEFAULT_GFXSTREAM_CAPSET_MASK),
            },
        }
    }